AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift

```diff
@@ -600,42 +617,45 @@ extension AWSMobileClient {
-        self.internalCredentialsProvider?.clearCredentials()
-        self.internalCredentialsProvider?.credentials()
+        return postSignInKeychainAndCredentialsUpdate(provider: .userPools)
     }
     
-    internal func performHostedUISuccessfulSignInTasks(disableFederation: Bool = false, session: AWSCognitoAuthUserSession, federationToken: String, federationProviderIdentifier: String? = nil, signInInfo: inout [String: String]) {
+    internal func performHostedUISuccessfulSignInTasks(disableFederation: Bool = false,
+                                                       session: AWSCognitoAuthUserSession,
+                                                       federationToken: String,
+                                                       federationProviderIdentifier: String? = nil,
+                                                       signInInfo: inout [String: String]) -> AWSTask<AWSCredentials>  {
         federationDisabled = disableFederation
         if federationProviderIdentifier == nil {
             self.cachedLoginsMap = [self.userPoolClient!.identityProviderName: federationToken]
         } else {
             self.cachedLoginsMap = [federationProviderIdentifier!: federationToken]
         }
-        self.mobileClientStatusChanged(userState: .signedIn, additionalInfo: signInInfo)
-        self.saveLoginsMapInKeychain()
-        self.setLoginProviderMetadataAndSaveInKeychain(provider: .hostedUI)
-        self.internalCredentialsProvider?.clearCredentials()
-        self.internalCredentialsProvider?.credentials()
+        return postSignInKeychainAndCredentialsUpdate(provider: .hostedUI)
     }
     
-    internal func performFederatedSignInTasks(provider: String, token: String) {
+    internal func performFederatedSignInTasks(provider: String, token: String) -> AWSTask<AWSCredentials>  {
         self.cachedLoginsMap = [provider:token]
-        self.mobileClientStatusChanged(userState: .signedIn, additionalInfo: [ProviderKey:provider, TokenKey: token])
         self.federationProvider = .oidcFederation
+        return postSignInKeychainAndCredentialsUpdate(provider: .oidcFederation)
+    }
+    
+    /// Post signin operations. Saves the login maps and provider metadata in keychain. Updates the credentials.
+    /// If credential update is successful, informs the listener with UserState.signedIn status.
+    ///
+    /// - Parameters
+    ///     - provider: provider to be updated in keychain.
+    func postSignInKeychainAndCredentialsUpdate(provider: FederationProvider) -> AWSTask<AWSCredentials> {
         self.saveLoginsMapInKeychain()
-        self.setLoginProviderMetadataAndSaveInKeychain(provider: .oidcFederation)
+        self.setLoginProviderMetadataAndSaveInKeychain(provider: provider)
         self.internalCredentialsProvider?.clearCredentials()
-        self.internalCredentialsProvider?.credentials()
+        return self.internalCredentialsProvider!.credentials();
```

Comment:
Instead of force-unwrapping, do a `guard let` here and return a TaskWithError if `internalCredentialsProvider` turns out to be `nil`

Reply:
Updated